### PR TITLE
Make useContextSelector concurrent mode safe

### PR DESCRIPTION
**Pull request by defunctzombie**
**User-Facing Changes**
None

**Description**
useContextSelector had side-effects in render. This is unsafe with react concurrent mode. This change removes the side-effects and simplifies the logic for updating the selected value.

### CODE REVIEW

**Comment by defunctzombie** [comment]
@jtbandes before I go and update all the tests related to this change - do you think this is the right change to make? It looks sensible to me.

**Comment by jtbandes** [comment]
The change makes sense, but the previous implementation was intentionally designed to not require 2 renders for any update (which is why the tests are failing). Do you think there is some way we can keep that behavior while being concurrent safe? Maybe it's just not possible without https://github.com/reactjs/rfcs/pull/119 ? Or I wonder if they have solved this behavior in https://github.com/dai-shi/use-context-selector? (Haven't explored this library much)

File: packages/studio-base/src/hooks/useContextSelector.ts [hunk 1]
```diff
@@ -36,19 +36,29 @@ export default function useContextSelector<T, U>(
...
+    if (shouldUpdate) {
+      state.current = {
+        contextValue,
+        selectedValue,
+        selector,
+      };
+    }
+  });
 
-  // Subscribe to context updates, and trigger a re-render when the selected value changes.
```
Comment (jtbandes) [inline packages/studio-base/src/hooks/useContextSelector.ts:51]: keep this comment?

**Review by jtbandes — APPROVED** [review]
This makes sense and feels like a more minimal change.

File: packages/studio-base/src/hooks/useContextSelector.ts [hunk 2]
```diff
@@ -36,19 +36,29 @@ export default function useContextSelector<T, U>(
...
   useLayoutEffect(() => {
```
Comment (defunctzombie) [inline packages/studio-base/src/hooks/useContextSelector.ts:62]:
```suggestion
  // Subscribe to context updates, and trigger a re-render when the selected value changes.
  useLayoutEffect(() => {
```

**Pull request merged**